Source/WebCore/Modules/cache/WorkerCacheStorageConnection.cpp

```diff
@@ -229,6 +231,20 @@ void WorkerCacheStorageConnection::dereference(DOMCacheIdentifier cacheIdentifie
     });
 }
 
+void WorkerCacheStorageConnection::lockStorage(const ClientOrigin& origin)
+{
+    callOnMainThread([mainThreadConnection = m_mainThreadConnection, origin = origin.isolatedCopy()]() {
```

Comment:
It seems there could still be race if LockCacheStorage is a separate message from CacheStorageAllCaches (e.g. network process might clear in-memory storage after it receives CacheStorageAllCaches and before LockCacheStorage). 
How about making network process automatically ref all caches for the web process connection when it receives CacheStorageAllCaches, and make web process send a DidRetrieveAllCaches/DereferenceCaches message to tell network process to remove ref to all caches (or caches it does not need to connect) in CacheStorageAllCaches reply? Web process could skip the cache identifier that is used to create new DOMCache.

This might be applied to CacheStorageOpenCache: network process ref the cache when receiving the message, and web process should send CancelOpenCache/DereferenceCache to tell network process to deref the cache when it decides to not create a new DOMCache with identifier or the DOMCache is destroyed.

Reply:
LockCacheStorage message happens before CacheStorageAllCaches message so there should be no race.
I thought about doing your approach but it seemed to me less readable (we would need to pass a callback to be executed on either worker or main thread as part of the CacheStorageAllCaches result callback).

Reply:
ok, then let's go with your implementation then